Self-referencing KeyPath falsely triggers self_in_property_initialization

### New Issue Checklist

- [X] I've Updated SwiftLint to the latest version.
- [X] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

When using a key path within property initialization (either as a parameter in a function/initializer for another type or as the property itself) I believe SwiftLint is falsely identifying the key path as referencing a function on `NSObject`.

```swift
final class NotActuallyReferencingSelf {

    // This doesn't actually reference `self` but it shows an error "`self` refers to the unapplied `NSObject.self()`"
    let keyPath: Any = \String.self

    // This does *not* show an error.
    let someType: Any = String.self
}
```

The error mentions `NSObject`, but `String` is a `struct` and you can verify this with any type `\Foo.self`.

### Command/Output

```bash
$ swiftlint --strict
```

> /Path/to/the/file/NotActuallyReferencingSelf.swift:10:5: error: Self in Property Initialization Violation: `self` refers to the unapplied `NSObject.self()` method, which is likely not expected; make the variable `lazy` to be able to refer to the current instance or use `ClassName.self` (self_in_property_initialization)

In reference to this case I believe we are using "`ClassName.self`".

### Environment

* SwiftLint version: 0.57.0
* Xcode version: 16.0.0
* Installation method used: Homebrew
* Configuration file: default

Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? If so, paste their
relative paths and respective contents.

- No nested config

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,6 +23,11 @@
   [SimplyDanny](https://github.com/SimplyDanny)
   [#5787](https://github.com/realm/SwiftLint/issues/5787)
 
+* Do not trigger `self_in_property_initialization` rule on `self` in
+  key paths expressions.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#5835](https://github.com/realm/SwiftLint/issues/5835)
+
 * Do not throw deprecation warning if deprecated property is not
   presented in configuration.  
   [chipp](https://github.com/chipp)

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/SelfInPropertyInitializationRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/SelfInPropertyInitializationRule.swift
@@ -68,6 +68,12 @@ struct SelfInPropertyInitializationRule: Rule {
                 func calculateB() -> String { "B" }
             }
             """, excludeFromDocumentation: true),
+            Example("""
+            final class NotActuallyReferencingSelf {
+                let keyPath: Any = \\String.self
+                let someType: Any = String.self
+            }
+            """, excludeFromDocumentation: true),
         ],
         triggeringExamples: [
             Example("""
@@ -102,7 +108,7 @@ private extension SelfInPropertyInitializationRule {
                 return
             }
 
-            let visitor = IdentifierUsageVisitor(identifier: .keyword(.self))
+            let visitor = IdentifierUsageVisitor(viewMode: .sourceAccurate)
             for binding in node.bindings {
                 guard let initializer = binding.initializer,
                       visitor.walk(tree: initializer.value, handler: \.isTokenUsed) else {
@@ -116,16 +122,12 @@ private extension SelfInPropertyInitializationRule {
 }
 
 private final class IdentifierUsageVisitor: SyntaxVisitor {
-    let identifier: TokenKind
     private(set) var isTokenUsed = false
 
-    init(identifier: TokenKind) {
-        self.identifier = identifier
-        super.init(viewMode: .sourceAccurate)
-    }
-
     override func visitPost(_ node: DeclReferenceExprSyntax) {
-        if node.baseName.tokenKind == identifier, node.keyPathInParent != \MemberAccessExprSyntax.declName {
+        if node.baseName.tokenKind == .keyword(.self),
+           node.keyPathInParent != \MemberAccessExprSyntax.declName,
+           node.keyPathInParent != \KeyPathPropertyComponentSyntax.declName {
             isTokenUsed = true
         }
     }

